src/main.rs

```diff
@@ -10,14 +10,51 @@
 
 mod cli;
 
-use std::process::ExitCode;
+use std::{io::Write, process::ExitCode};
 
 use bft_interp::VirtualMachine;
 use bft_types::BfProgram;
 use clap::Parser;
+use std::io::{stdin, stdout};
 
 use cli::Args;
 
+struct WriterWithTrailingNewline<T: Write> {
+    inner_writer: T,
+    last_byte: u8,
+}
+
+impl<T: Write> WriterWithTrailingNewline<T> {
+    fn new(inner_writer: T) -> Self {
+        Self {
+            inner_writer,
+            last_byte: 0,
+        }
+    }
+}
+
+impl<T: Write> Write for WriterWithTrailingNewline<T> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        if let Some(last_byte) = buf.last() {
+            self.last_byte = *last_byte;
+        }
+        self.inner_writer.write(buf)
```

Comment:
This isn't going to guarantee to write everything - either use write_all or else use write, but then get the nbytes from that to decide which to check for newlineness in your buffer.  See me if you can't work out how to do these.